src/Microsoft.Health.Fhir.TemplateManagement/OCIFileManager.cs

```diff
@@ -58,34 +88,59 @@ public void PackOCIImage(bool ignoreBaseLayers = false)
                 baseArtifactLayers = _overlayFS.ReadBaseLayers();
             }
 
-            var snapshotLayer = GenerateBaseFileLayer(baseArtifactLayers);
+            var baseOCIFileLayers = SortBaseLayers(baseArtifactLayers);
+            var snapshotLayer = _overlayOperator.MergeOCIFileLayers(baseOCIFileLayers);
             var diffLayer = _overlayOperator.GenerateDiffLayer(mergedLayer, snapshotLayer);
             var diffArtifactLayer = _overlayOperator.ArchiveOCIFileLayer(diffLayer);
-            if (baseArtifactLayers.Count == 0)
-            {
-                _overlayFS.WriteBaseLayers(new List<OCIArtifactLayer> { diffArtifactLayer });
-            }
-
-            baseArtifactLayers.Add(diffArtifactLayer);
-            _overlayFS.WriteImageLayers(baseArtifactLayers);
+            baseOCIFileLayers.Add((OCIFileLayer)diffArtifactLayer);
+            var sortedlayers = baseOCIFileLayers.Where(layer => layer.Content != null).Select(layer => (OCIArtifactLayer)layer).ToList();
+            return sortedlayers;
         }
 
-        public async Task PushOCIImageAsync()
+        public async Task PushOCIImageAsync(List<OCIArtifactLayer> sortedlayers)
         {
-            await _orasClient.PushImageAsync(_overlayFS.WorkingImageLayerFolder);
+            _overlayFS.WriteImageLayers(sortedlayers);
+            await _orasClient.PushImageAsync(_overlayFS.WorkingImageLayerFolder, sortedlayers.Select(layer => layer.FileName).ToList());
         }
 
-        private OCIFileLayer GenerateBaseFileLayer(List<OCIArtifactLayer> baseArtifactLayers)
+        private List<OCIFileLayer> SortBaseLayers(List<OCIArtifactLayer> baseArtifactLayers)
         {
-            if (baseArtifactLayers == null || baseArtifactLayers.Count == 0)
+            List<OCIFileLayer> sortedBaseLayers = new List<OCIFileLayer>();
+            if (baseArtifactLayers.Count == 0)
             {
-                return null;
+                return sortedBaseLayers;
```

Comment:
I see a lot of cases returning null. What's the difference in returning null or empty list?

Reply:
Empty list is more safe for the list operations(such as add). return null if the type is only an object.